mesonbuild/rewriter.py

```diff
@@ -685,6 +690,17 @@ def find_node(src):
             }
             self.add_info('target', target['id'], test_data)
 
+        # Sort files
+        for i in to_sort_nodes:
+            convert = lambda text: int(text) if text.isdigit() else text.lower()
+            alphanum_key = lambda key: [convert(c) for c in re.split('([0-9]+)', key)]
+            path_sorter = lambda key: (0 if '/' in key else 1, key.split('/'), alphanum_key(key))
```

Comment:
This looks fairly complicated to explain. I'm not sure if this is worth having possibly better sorting. After all users not using this integration have to be able to produce the same ordering easily.
Maybe we should just go with simple sorting that is also available in many editors or IDEs.

Reply:
What it does is sorting all directories first, then it sorts by path. It also handles natural order sorting:

```
'aaa/a/a1.c',
'aaa/b/b1.c',
'aaa/b/b2.c',
'aaa/f1.c',
'aaa/f2.c',
'aaa/f3.c',
'bbb/a/b1.c',
'bbb/b/b2.c',
'bbb/b3.c',
'bbb/b4.c',
'bbb/b5.c',
'a1.c',
'a10.c',
'a100.c',
'a101.c',
'a110.c',
'a2.c',
'a20.c',
'a210.c',
'a3.c',
'a30.c',
'a666.c',
'b1.c',
'c2.c'
```

While automatic sorting by IDE's may indeed pose a problem, inserting new files by hand should be fairly easy. Also, @jpakkane wanted this feature :)

> The only thing that comes to mind is that it should sort things "logically" rather than asciibetically. That is it should not do this:
> 
> ```
> aaa.c
> aab/aab.c
> aab/aac.c
> aad.c
> ```
> 
> but instead this:
> 
> ```
> aab/aab.c
> aab/aaa.c
> aaa.c
> aad.c
> ```
> 
> And probably also [natural order sorting](https://blog.codinghorror.com/sorting-for-humans-natural-sort-order/) (we already do that for test case ordering).

The problem with the "stupid" ASCII sorting is that some projects may prefer a more sophisticated sorting algorithm (like this one).
So, the question is whether we enforce the stupid ASCII sorting, or a "smarter" sorting algorithm. However, in both cases, inserting new files by hand should be fairly trivial for humans to do.

I want to add support for a rewriter config file for the next release anyway (similar to `.clang-format`), where this behavior can be configured.